File: Final_Project/Code/Project_4/Bai_toan_5/Exercises/10/CSS_bt5_ex10.cpp
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int value;         // gia tri cua nut
    int first_child;   // chi so cua con dau tien
    int next_sibling;  // chi so cua anh em ke tiep
    int parent;        // chi so cua cha
    int num_children;  // so con truc tiep

    TreeNode(int val)
      : value(val), first_child(-1),
        next_sibling(-1), parent(-1),
        num_children(0) {}
};

class Tree {
public:
    int n;                  // so nut toi da
    vector<TreeNode*> nodes;// mang luu cac nut
    int root_node;          // chi so nut goc
    int next_idx;           // chi so tiep theo de cap phat

    Tree(int vertices)
      : n(vertices), root_node(-1), next_idx(0) {
        nodes.resize(n, nullptr);
    }

    void set_root(int value) {
        // thiet lap nut goc
        nodes[0] = new TreeNode(value);
        root_node = 0;
        next_idx = 1;
    }

    void add_child(int parent_idx, int value) {
        // them con vao nut parent_idx
        if (next_idx >= n) throw runtime_error("Tree is full");
        int new_node_idx = next_idx;
        nodes[new_node_idx] = new TreeNode(value);
        nodes[new_node_idx]->parent = parent_idx;
        if (nodes[parent_idx]->first_child == -1) {
            nodes[parent_idx]->first_child = new_node_idx;
        } else {
            int current = nodes[parent_idx]->first_child;
            while (nodes[current]->next_sibling != -1)
                current = nodes[current]->next_sibling;
            nodes[current]->next_sibling = new_node_idx;
        }
        nodes[parent_idx]->num_children++;
        next_idx++;
    }

    int parent(int v) { return nodes[v]->parent; }
    int first_child(int v) { return nodes[v]->first_child; }
    int next_sibling(int v){ return nodes[v]->next_sibling; }
    int number_of_children(int v){
        return nodes[v]->num_children;
    }
};

int main() {
    Tree t(5);
    t.set_root(0);
    t.add_child(0, 1);
    t.add_child(0, 2);
    t.add_child(0, 3);
    cout << "Number of children of root: "
         << t.number_of_children(0) << endl; // 3
    cout << "First child of root: "
         << t.first_child(0) << endl;         // 1
    cout << "Next sibling of first child: "
         << t.next_sibling(1) << endl;        // 2
    cout << "Parent of node 2: "
         << t.parent(2) << endl;             // 0
    return 0;
}

File: Final_Project/Code/Project_4/Bai_toan_5/Exercises/10/CSS_bt5_ex10.py
class TreeNode:
    def __init__(self, value):
        # value: gia tri cua nut
        # first_child: chi so cua con dau tien
        # next_sibling: chi so cua anh em ke tiep
        # parent: chi so cua cha
        # num_children: so con truc tiep
        self.value = value
        self.first_child = None
        self.next_sibling = None
        self.parent = None
        self.num_children = 0

class Tree:
    def __init__(self, n):
        # n: so nut toi da
        # nodes: mang luu cac nut
        # root_node: chi so cua nut goc
        # next_idx: chi so tiep theo de cap phat
        self.n = n
        self.nodes = [None] * n
        self.root_node = None
        self.next_idx = 0

    def set_root(self, value):
        # thiet lap nut goc
        self.nodes[0] = TreeNode(value)
        self.root_node = 0
        self.next_idx = 1

    def add_child(self, parent_idx, value):
        # them con vao nut parent_idx
        if self.next_idx >= self.n:
            raise ValueError("Tree is full")
        new_node_idx = self.next_idx
        self.nodes[new_node_idx] = TreeNode(value)
        self.nodes[new_node_idx].parent = parent_idx
        if self.nodes[parent_idx].first_child is None:
            self.nodes[parent_idx].first_child = new_node_idx
        else:
            current = self.nodes[parent_idx].first_child
            while self.nodes[current].next_sibling is not None:
                current = self.nodes[current].next_sibling
            self.nodes[current].next_sibling = new_node_idx
        self.nodes[parent_idx].num_children += 1
        self.next_idx += 1

    def parent(self, v):
        return self.nodes[v].parent

    def first_child(self, v):
        return self.nodes[v].first_child

    def next_sibling(self, v):
        return self.nodes[v].next_sibling

    def number_of_children(self, v):
        return self.nodes[v].num_children

# Vi du su dung
if __name__ == "__main__":
    t = Tree(5)
    t.set_root(0)
    t.add_child(0, 1)
    t.add_child(0, 2)
    t.add_child(0, 3)
    print("Number of children of root:", t.number_of_children(0))  # 3
    print("First child of root:", t.first_child(0))                # 1
    print("Next sibling of first child:", t.next_sibling(1))       # 2
    print("Parent of node 2:", t.parent(2))                         # 0